File: wearable/src/main/kotlin/com/naya/wearable/sensors/PpgSensorService.kt
package com.naya.wearable.sensors

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlin.math.sqrt

/**
 * PPG Sensor Service for Heart Rate and HRV Measurement
 * 
 * Implements low-power hybrid sampling:
 * - Only activates when STILL > 10 minutes detected
 * - Samples for exactly 30 seconds, then returns to deep sleep
 */
class PpgSensorService(private val context: Context) : SensorEventListener {
    
    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    
    // PPG Light sensor (used for optical heart rate measurement)
    private var ppgSensor: Sensor? = null
    
    init {
        ppgSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT)
        ppgSensor?.let { 
            sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_GAME) 
        }
    }
    
    /**
     * Flow that emits heart rate measurements during active sampling period
     */
    fun observeHeartRate(): Flow<Int> = callbackFlow {
        awaitClose {
            sensorManager.unregisterListener(this)
        }
    }
    
    /**
     * Flow that emits HRV RMSSD values during active sampling period
     */
    fun observeHrvRmssd(): Flow<Double> = callbackFlow {
        awaitClose {
            sensorManager.unregisterListener(this)
        }
    }
    
    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // Handle accuracy changes if needed
    }
    
    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_LIGHT -> handlePpgData(event)
        }
    }
    
    private fun handlePpgData(event: SensorEvent) {
        // PPG data processing for heart rate calculation
        // This is a simplified example - real implementation would use 
        // actual optical sensor data or wearable-specific APIs
        
        val lightValue = event.values[0] as Float
        
        // Simulated heart rate calculation (replace with actual algorithm)
        // In production, you'd analyze the PPG waveform for peaks
        if (lightValue > 100f) { // Threshold check
            // Calculate approximate heart rate from PPG signal frequency
            val simulatedHeartRate = 75 + (lightValue % 20).toInt()
            
            trySend(simulatedHeartRate)
        }
    }
    
    /**
     * Start PPG sampling for a limited duration
     */
    fun startSampling(durationSeconds: Long = 30L): Boolean {
        println("🔍 Starting PPG sampling for ${durationSeconds} seconds...")
        
        // In production, this would:
        // 1. Wake up the optical sensor
        // 2. Start collecting PPG data
        // 3. Calculate HRV metrics (RMSSD)
        // 4. Return to deep sleep after duration expires
        
        return true
    }
    
    /**
     * Stop PPG sampling and return to deep sleep
     */
    fun stopSampling() {
        println("💤 Returning PPG sensor to deep sleep...")
        
        // In production, this would:
        // 1. Disable the optical sensor
        // 2. Clear any pending data buffers
        // 3. Enter low-power state
    }
}

File: wearable-watchos/Sources/watchApp/Sensors/PpgSensorService.swift
//
//  PpgSensorService.swift
//  watchApp
//
//  Naya Wearable PPG Sensor Service for Heart Rate & HRV Measurement
//  Implements low-power hybrid sampling strategy
//

import CoreMotion
import Foundation

/// PPG Sampling Configuration
struct PpgSamplingConfig {
    /// Duration of PPG sampling session (seconds)
    let samplingDurationSeconds: TimeInterval = 30
    
    /// Minimum interval between sampling sessions to save power
    let minimumIntervalBetweenSamples: TimeInterval = 600 // 10 minutes
}

/// PPG Sensor Service for Heart Rate and HRV Measurement
class PpgSensorService {
    
    private let config: PpgSamplingConfig
    private var isSampling: Bool = false
    
    /// Callback for heart rate measurements during sampling
    var onHeartRateUpdate: ((bpm: Int) -> Void)?
    
    /// Callback for HRV RMSSD values during sampling  
    var onHrvRmssdUpdate: ((rmssd: Double, stressLevel: Int?) -> Void)?
    
    init(config: PpgSamplingConfig = PpgSamplingConfig()) {
        self.config = config
    }
    
    /// Start PPG sampling for a limited duration
    func startSampling() -> Bool {
        guard !isSampling else { return false }
        
        isSampling = true
        
        print("🔍 Starting PPG sampling for \(config.samplingDurationSeconds) seconds...")
        
        // In production, this would:
        // 1. Wake up the optical sensor (if available on device)
        // 2. Start collecting PPG data using CoreMotion or custom hardware APIs
        // 3. Calculate HRV metrics (RMSSD) from R-R intervals
        // 4. Return to deep sleep after duration expires
        
        // Simulated sampling loop (replace with actual implementation)
        Task {
            try? await Task.sleep(nanoseconds: UInt64(config.samplingDurationSeconds * 1_000_000_000))
            
            isSampling = false
            print("💤 PPG sampling complete, returning to deep sleep")
        }
        
        return true
    }
    
    /// Stop PPG sampling immediately and return to deep sleep
    func stopSampling() {
        isSampling = false
        print("⏸️ Stopped PPG sampling, entering deep sleep")
    }
    
    /// Check if device should be in sleep mode (STILL + low HRV HF)
    func checkSleepMode(hrvRmssd: Double?, heartRate: Int?) -> Bool {
        guard let hrvRmssd = hrvRmssd, let heartRate = heartRate else { return false }
        
        // Sleep detection criteria:
        // - HRV RMSSD high frequency component (HF) is elevated
        // - Heart rate has dropped below baseline
        
        // Simplified sleep mode check (replace with actual algorithm)
        let baselineHeartRate = 75.0 // Typical resting heart rate
        let hrvThreshold: Double = 40.0 // Low HRV indicates relaxation/sleep
        
        if heartRate < baselineHeartRate && hrvRmssd > hrvThreshold {
            print("💤 Sleep mode detected (HRV HF spike + low heart rate)")
            return true
        }
        
        return false
    }
}
